### ISSUE TITLE
Ajax and Jquery autocomplete response slow speed

### ISSUE THREAD

**Issue by Vibrat** [opencart/opencart#6469]
I'm digging in filter-product feature in admin page which returns suggestions everytime we click or typing in the input. The current technique is to use Jquery 'autocomplete' and Ajax. It seems nice to understand but the speed is slow. is there any way to speed up this.

**Controller:** 
admin/index.php?route=catalog/product

**View file**
admin\view\template\catalog\product_list.twig

`$('input[name=\'filter_name\']').autocomplete({
	'source': function(request, response) {
		$.ajax({
			url: 'index.php?route=catalog/product/autocomplete&user_token={{ user_token }}&filter_name=' +  encodeURIComponent(request),
			dataType: 'json',
			success: function(json) {
				response($.map(json, function(item) {
					return {
						label: item['name'],
						value: item['product_id']
					}
				}));
			}
		});
	},
	'select': function(item) {
		$('input[name=\'filter_name\']').val(item['label']);
	}
});`

**Comment by stalker780** [comment]
Perhaps autocomplete must have a 300-400ms delay after keyup.

Now it calls ajax right after each keypress.
For example word "product" calls 7 ajax requests almost simultaneously.

If you have a lot of products and slow hosting this can give noticable server response time.

**Comment by lucasjkr** [comment]
Per my count, each Ajax request is generating 334 queries.

And that's with a tiny test instance with the following record counts:

oc_product: 118 records
oc_product_description: 118 records
oc_option: 11 records
oc_product_option: 328 records

 etc... 

I'm thinking improving this situation could require a new function in /admin/model/catalog/product specifically for autocompletes. Granted I've spent literally 5 minutes looking (more time typing this), so I might be rushing to ideas prematurely.

Nonetheless... loading the page causes **44** queries to run, which is fine. Clicking in the search box which causes the auto complete drop down to run, brings the query count to **211**. Typing 'a' brings the query count to **545**! Typing 'p' after that brings it to **879**.

Literally, each character typed is causing **334** queries to occur. This is going to hammer any system with more than a few products and/or more than a few simultaneous users.

So, in your 7 AJAX call scenario, there are 2,338 queries executing. That hurts.

I think we need to put our brains together to address that.

**Comment by Vibrat** [comment]
@lucasjkr Here is what i have been so far. This technique uses Ajax, LocalStorage and HTML5 tag 'datalist' to deal with this problem. What it helps include:

- Loading LocalStorage to get results of previous search everytime user refresh or access to the page.
- Ajax calls and gets data from Database, then use javascript to check whether if it's not included in the datalist, if not then adding it to the top of the datalist tag.
- Save datalist to localStorage everytime Ajax is finished.
- Prevent multiple trigger Ajax (setTimeOut to 500ms to limit numbers of AJAX query)
- By doing this we can actually get data of user search behaviors.

```
    var typingTimer;                //timer identifier
    var doneTypingInterval = 500;  //time in ms, 5 second for example


    //on keyup, start the countdown
    $('input[name=\'filter_name\']').keypress(function(){
        clearTimeout(typingTimer);
        if ($(this).val) {
            typingTimer = setTimeout(function(){
                //do stuff here e.g ajax call etc....
                    $.ajax({
                       url: 'index.php?route=catalog/product/autocompletename&user_token={{ user_token }}&filter_name=' +  encodeURIComponent($('input[name=\'filter_name\']').val()),
                       dataType: 'json',
                       success: function(json) {
                           for (key in json) {
                                datalist  = document.getElementById('input-product-name');
                                options   = datalist.getElementsByTagName('option');
                                dataexist = false;

                                for (id in options) {
                                    if (options[id].value === json[key].name) {
                                        dataexist = true;
                                    }
                                }
                                if (dataexist === false) {
                                    datalist.insertAdjacentHTML('afterbegin', '<option value="'+ json[key].name + '"></option>');

                                    if (typeof(Storage) !== "undefined") {
                                        // Store
                                        localStorage.setItem("product-name-datalist", datalist.innerHTML);
                                    }
                                }
                           }
                       }
                    });
            }, doneTypingInterval);
        }
    });
```

**Comment by lucasjkr** [comment]
The issue isn't the AJAX and the delay, it's the volume of data Opencart is sifting through.

If you search in the product name field, then for every key you type, it queries the product description table for records with a product name in the current language starting with whichever string you entered, and then gets all the option information for each product it returned.

On your site, authenticate as Admin and then try going to this URL, copy your user token and replace it in the the URL below (also replace the domain name and port, obviously):

http://opencart.local:8080/admin/index.php?route=catalog/product/autocomplete&filter_name=a&user_token={{your user token}}

This is the problem. 

We need a saner approach to autocomplete. 

All we're returning to the user is the product ID and the field they were querying (Product Name, Model, etc). No reason to bog down everything with all the option data. If that stuff is needed by other features, that's fine. Just need to redo the autocomplete and/or make a new function in the model to only return the information we need, and not waste time searching fields that we have no need for. 

That's the solution, plain and simple.

**Comment by Vibrat** [comment]
Selecting only neccessary information  from database does help in speed. There is still problems with speed in Ajax but it seems better.  Combining two choices of frontend and backend handling would be a nice choice for me.

**Comment by danielkerr** [comment]
this has been fixed already using js and adding a 

if you need a fix just use the code below "// Autocomplete" in common.js master branch
